tests/validation/tests/v3_api/test_websocket.py

```diff
@@ -0,0 +1,154 @@
+    target = last_message.split("\r\n")[1]
+    assert "daemonset.apps/nginx-ingress-controller" in target
+    last_message = ''
+
+    ws.close()
+
+
+def test_websocket_exec_shell():
+    global last_message
+    url = 'wss://' + CATTLE_TEST_URL[8:] + \
+        '/k8s/clusters/' + namespace["cluster"].id + \
+        '/api/v1/namespaces/' + namespace["ns"] + \
+        '/pods/' + namespace["pod"].name + \
+        '/exec?container=' + namespace["pod"].containers[0].name + \
+        '&stdout=1&stdin=1&stderr=1&tty=1&command=%2Fbin%2Fsh&command=-c' \
+        '&command=TERM%3Dxterm-256color%3B%20export%20TERM%3B%20%5B%20-x%20' \
+        '%2Fbin%2Fbash%20%5D%20%26%26%20(' \
+        '%5B%20-x%20%2Fusr%2Fbin%2Fscript%20%5D%20%26%26%20%2Fusr%2Fbin' \
+        '%2Fscript%20-q%20-c%20%22%2Fbin%2Fbash%22%20%2Fdev%2Fnull%20%7C%7C' \
+        '%20exec%20%2Fbin%2Fbash)%20%7C%7C%20exec%20%2Fbin%2Fsh'
+
+    ws = websocket.create_connection(
+        url=url,
+        sslopt={"cert_reqs": ssl.CERT_NONE},
+        subprotocols=["base64.channel.k8s.io"],
+        timeout=30,
+        cookie="R_SESS=" + USER_TOKEN
+    )
+    assert ws.connected, "failed to build the websocket"
+
+    thread = Thread(target=receiver, args=(ws, True))
+    thread.daemon = True
+    thread.start()
+    time.sleep(1)
+
+    cmd = "pwd"
+    send_a_command(ws, cmd)
+    print('shell command and output:\n' + last_message + '\n')
+    assert '/etc/nginx' in last_message, "failed to run the kubectl command"
+    last_message = ''
+
+    ws.close()
+
+
+def test_websocket_view_logs():
+    global last_message
+    url = 'wss://' + CATTLE_TEST_URL[8:] + \
+        '/k8s/clusters/' + namespace["cluster"].id + \
+        '/api/v1/namespaces/' + namespace["ns"] + \
+        '/pods/' + namespace["pod"].name + \
+        '/log?container=' + namespace["pod"].containers[0].name + \
+        '&tailLines=500&follow=true&timestamps=true&previous=false'
+    ws = websocket.create_connection(
+        url=url,
+        sslopt={"cert_reqs": ssl.CERT_NONE},
+        subprotocols=["base64.binary.k8s.io"],
+        timeout=30,
+        cookie="R_SESS=" + USER_TOKEN
+    )
+    assert ws.connected, "failed to build the websocket"
+
+    thread = Thread(target=receiver, args=(ws, False))
+    thread.daemon = True
+    thread.start()
+    time.sleep(1)
+    print('shell command and output:\n' + last_message + '\n')
+    assert 'NGINX Ingress controller' in last_message, "failed to view logs"
+    last_message = ''
+
+    ws.close()
+
+
+def receiver(socket, skip):
+    global last_message
+    while True:
+        try:
+            data = socket.recv()
+            if skip:
+                data = data[1:]
+            if len(data) < 5:
+                pass
+            data = base64.b64decode(data).decode()
+            last_message += data
+        except websocket.WebSocketConnectionClosedException:
+            print("Connection closed")
+            break
+        except ssl.SSLZeroReturnError:
```

Comment:
When does this exception get raised?

Reply:
I used to see this error but can not reproduce it anymore, so the exception is removed.